src/System.IO.Packaging/src/System/IO/Packaging/PackUriHelper.cs

```diff
@@ -719,21 +712,30 @@ private static void EscapeAsciiChar(char ch, char[] to, ref int pos)
         /// </summary>
         /// <param name="path"></param>
         /// <returns></returns>
-        private static string EscapeSpecialCharacters(string path)
+        private unsafe static string EscapeSpecialCharacters(string path)
         {
-            string characterString;
+            var builder = new StringBuilder(path.Length);
+            const int charsLength = 3;
+            char* chars = stackalloc char[charsLength];
 
             // Escaping for the following - '%'; '@'; ',' and '?'
             // !!Important!! - The order is important - The '%' sign should be escaped first.
             // This is currently enforced by the order of characters in the _specialCharacters array
             foreach (char c in s_specialCharacters)
             {
-                characterString = c.ToString();
-                if (path.Contains(characterString))
-                    path = path.Replace(characterString, HexEscape(c));
+                if (path.IndexOf(c) < 0)
+                {
+                    builder.Append(c);
+                }
+                else
+                {
+                    int pos = 0;
+                    EscapeAsciiChar(c, chars, ref pos);
+                    builder.Append(chars, charsLength);
+                }
```

Comment:
This isn't a valid transformation.  It does not provide the same functionality as the original.  And the fact that it's passing tests worries me such that we should ensure a test is added that covers this code.